lib/statsd_middleware.rb

```diff
@@ -95,6 +107,18 @@ def call(env)
   def get_source_app(env)
     source_app = env['HTTP_SOURCE_APP_NAME']
 
+    # When the source app name is not provided through the header, attempt to
+    # match the referrer to the root URL of one of the whitelisted apps.
+    if source_app.nil?
+      referer = env['HTTP_REFERER']
+
+      if referer
+        *, source_app = SOURCE_APP_URL_MAP.find do |root_url, _|
+          URI.parse(referer).path.start_with? root_url
+        end
+      end
+    end
```

Comment:
I'm a little concerned at the performance of this code since it runs on every single request. One easy thing is to move 

```ruby
path = URI.parse(referer).path
```

outside the enumeration.

Reply:
Maybe some testing is in order to see how much time iterating over our entire `SOURCE_APP_URL_MAP` adds.

Reply:
It may run a lot from apps where the source_app isn't being sent as a header, but this will be avoided when it is provided.